core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java

```diff
@@ -4346,7 +4336,21 @@ private void convertSelectList(
 
     relBuilder.push(bb.root())
         .projectNamed(exprs, fieldNames, true);
-    bb.setRoot(relBuilder.build(), false);
+
+    RelNode project = relBuilder.build();
+
+    final RelNode r;
+    final CorrelationUse p = getCorrelationUse(bb, project);
+    if (p != null) {
+      assert p.r instanceof Project;
+      Project proj = (Project) p.r;
+      r = LogicalProject.create(proj.getInput(), proj.getHints(),
```

Comment:
I've seen that this pattern appears in other places in `SqlToRelConverter` but it is a bit weird that from one hand we are asserting that the project is an instance of a superclass (`Project`) and on the other hand we are creating a `LogicalProject`. Can we use the `relbuilder` to create the new project? Would it be better to rely on `project#copy`?

Reply:
I think this assertion primary used to avoid "Unchecked cast" warning throwing by a static code analyzers. Here we are converting sqlNode to a relation tree, so LogicalProject is a reasonable result. Parent class `Project` give us sufficent information to create a new `LogicalProject` with found set of correlated variables.

We could use the `relBuilder`, but it doesn't change a lot.

Reply:
The class allows somebody to pass their own customized `RelBuilder` which can create a `HiveProject` instead of `LogicalProject`. In various places the `SqlToRelConverter` creates explicitly `Logical` expressions instead passing via the `relBuilder` which currently makes the `RelBuilder` customization kind of useless but this may change in the near future.

Reply:
Oh, now I see. Then it definitely would be better to rewrite with relBuilder.

Reply:
Hi, @zabetak! I've reworked PR in order to use `relBuilder` to create a correlated project after deduplication. Could you take a look again please?